design/routing-design.md

```diff
@@ -123,6 +123,34 @@ Exact Path based routing makes decisions based upon the path of the request (e.g
 Wildcard path based routing allows for a portion of the path to be dynamic (e.g. /path/*/foo).
 This is useful to allow for many types of request paths to be routed to the same backed without specifying them explicitly up front.
 
+- Wildcards defined cannot be the end of the path (e.g. `/foo/*/bar` or `/*/foo` not `/foo*`)
```

Comment:
It would be really useful for me to have a condition type defined so we can write down these rules and play with them concisely. 

I don't want to call it wildcard, I think that's a bit specific, and would come with weird rules that the path _must_ contain some glob character ie, `wildcard: /api/` ? what does this match?

One idea might be `path: /api/*/foo/` where path is defined as an _exact match_. That is the path must start with `/api/` and end with /`foo` and have _something_ in between those two, that is it would not match `/api/foo/` or `/api//foo/`

Reply:
My thought was to not have a specific condition type. If we see a `*` in a prefix, then we build out a regex style match as a prefix match. If we implement exact later on, then we build out the regex that implements an exact match. 

I don't feel super strong about this stance, so happy to create more, it just adds to the spec and I don't feel that it buys us much. 

> don't want to call it wildcard, I think that's a bit specific
That's fine, any thoughts then on a name?

> path must contain some glob character ie, wildcard: /api/ ? what does this match?
To me that's an error. If you define a specific name for this type of matching, but don't provide the special character (i.e. `*`) then to me you entered an invalid value.

Reply:
Thanks for bearing with me, you've won me over that we could add `*` to `prefix:`. What I'd like to see is the introduction rewritten to something like this

## Wildcard matching

Certain condition types, for example `prefix` support wildcard style matching. The specifics depend on the condition type.

### Prefix condition wildcards

- Can have one or more `*` in prefix value.
- Cannot start or end with a `*` (starting with a `*` is prohibited by the general rule that `prefix:` must start with a `/`)
- Prefix conditions with wildcards are merged as per the general rule for merging prefix conditions.

When interpreting a prefix condition containing wildcards the following rules are applied. For the condition `prefix: /api/*/users` (either expressed directly, or formed via merging prefix conditions) the request's URL must meet these criteria:
1. It must start  with `/api/`
2. Following `/api/`  there must be one or more characters _BUT NOT_ `/users`. These are considered to be the wildcard characters.
3. Following the wildcard characters must be the string `/users`.

Specifically, wildcards match one or more characters; the condition `prefix: /api/*/users` _does not_ match the url `/api/users/foo`.